pkg/controller/route/route_controller.go

```diff
@@ -88,11 +89,46 @@ func New(routes cloudprovider.Routes, kubeClient clientset.Interface, nodeInform
 		clusterName:      clusterName,
 		clusterCIDR:      clusterCIDR,
 		nodeLister:       nodeInformer.Lister(),
+		nodeInformer:     nodeInformer.Informer(),
 		nodeListerSynced: nodeInformer.Informer().HasSynced,
 		broadcaster:      eventBroadcaster,
 		recorder:         recorder,
 	}
 
+	notified := func() {
+		if err := rc.reconcileNodeRoutes(); err != nil {
+			klog.Errorf("Couldn't reconcile node routes: %v when notified", err)
+		}
+	}
+
+	rc.nodeInformer.AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1.Node:
+					klog.Infof("Node %s event notified", t.Name)
+					return true
+				default:
+					return false
+				}
+			},
+
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					klog.Info("Add node, route reconcile")
+					notified()
+				},
+				UpdateFunc: func(oldObj, newObj interface{}) {
```

Comment:
Update events are actually really expensive because you get a node update event for every heart beat from the node (~ every 5 - 10 seconds per node). This might change as we use node leases for node heartbeats but until then this may be too expensive.

Reply:
I believe this was the original intention behind why routes controller does a full list of nodes over a fixed period. Please correct me if I'm wrong @bowei

Reply:
related: https://github.com/kubernetes/kubernetes/issues/60646

Reply:
It seems that `Update` function is redundant. In most cases, updates of nodes don't affect routes.